Sources/Amplitude/Plugins/iOS/UIKitUserInteractions.swift

```diff
@@ -0,0 +1,163 @@
+
+        guard
+            sendActionResult,
+            let view = sender as? UIView,
+            view.shouldTrack(action, for: event)
+        else { return sendActionResult }
+
+        let userInteractionEvent = view.eventFromData(with: NSStringFromSelector(action).components(separatedBy: ":").first ?? "")
+
+        UIKitUserInteractions.amplitudeInstances.allObjects.forEach {
+            $0.track(event: userInteractionEvent)
+        }
+
+        return sendActionResult
+    }
+}
+
+extension UIView {
+    private static let viewHierarchyDelimiter = " -> "
+
+    struct ViewData {
+        let viewController: String?
+        let title: String?
+        let accessibilityLabel: String?
+        let action: String
+        let targetViewClass: String
+        let targetText: String?
+        let hierarchy: String
+    }
+
+    func eventFromData(with action: String) -> UserInteractionEvent {
+        let viewData = extractData(with: action)
+        return UserInteractionEvent(
+            viewController: viewData.viewController,
+            title: viewData.title,
+            accessibilityLabel: viewData.accessibilityLabel,
+            action: viewData.action,
+            targetViewClass: viewData.targetViewClass,
+            targetText: viewData.targetText,
+            hierarchy: viewData.hierarchy)
+    }
+
+    func extractData(with action: String) -> ViewData {
+        let viewController = owningViewController
+        return ViewData(
+            viewController: viewController?.descriptiveTypeName,
+            title: viewController?.title,
+            accessibilityLabel: accessibilityLabel,
+            action: action,
+            targetViewClass: descriptiveTypeName,
+            targetText: title,
+            hierarchy: sequence(first: self, next: { $0.superview })
+                .map { $0.descriptiveTypeName }
+                .joined(separator: UIView.viewHierarchyDelimiter))
+    }
+}
+
+extension UIResponder {
+    var descriptiveTypeName: String {
+        String(describing: type(of: self))
+    }
+
+    var owningViewController: UIViewController? {
+        return self as? UIViewController ?? next?.owningViewController
+    }
+}
+
+extension UIView {
+    var title: String? {
+        switch self {
+        case let button as UIButton:
+            return button.currentTitle
+        case let segmentedControl as UISegmentedControl:
+            return segmentedControl.titleForSegment(at: segmentedControl.selectedSegmentIndex)
+        default:
+            return nil
+        }
+    }
+
+    func shouldTrack(_ action: Selector, for event: UIEvent?) -> Bool {
+        switch self {
+        case is UITextField:
+            return false
+        #if !os(tvOS)
+        case is UISlider:
+            return event?.allTouches?.contains { $0.phase == .ended && $0.view == self } ?? false
+        #endif
+        default:
+            return true
```

Comment:
Here, using protocol instead adds a lot of boilerplate code due this Swift bug: https://github.com/swiftlang/swift/issues/42725 
So I think this is a better way to add `title` and `shouldTrack` to each control.

Reply:
I'm bringing back the `Trackable` protocol, and adding `@objc` seems to make it work. Does it make sense to add the `amp_` prefix to `title` and `shouldTrack` since they're annotated with `@objc`?

Also, I think UIView is a good extension point for the `Trackable` protocol since the implementation checks whether the sender of the action method is a UIView before continuing.

Reply:
Yes, please add the amp_ prefix.

I agree that UIView makes more sense with the protocol approach, the current form seems relatively non-specific to view (as we're checking all the subtypes), and on larger projects it helps to keep extensions on common objects fairly minimal.